multibase/src/Data/ByteString/BaseN.hs

```diff
@@ -0,0 +1,932 @@
+type BaseIdentity      = AtBase "id"
+type Base16upper       = AtBase "16u"
+type Base32hex         = AtBase "16x"
+type Base32hexupper    = AtBase "16xu"
+type Base32hexpad      = AtBase "16xp"
+type Base32hexpadupper = AtBase "16xup"
+type Base32            = AtBase "32"
+type Base32upper       = AtBase "32u"
+type Base32pad         = AtBase "32p"
+type Base32padupper    = AtBase "32pu"
+type Base58flickr      = AtBase "58flickr"
+type Base58btc         = AtBase "58btc"
+type Base64pad         = AtBase "64p"
+type Base64url         = AtBase "64url"
+type Base64urlpad      = AtBase "16urlp"
+
+-- Compact ---------------------------------------------------------------------
+
+-- | A more memory-efficient representation of base-n encoded bytes.
+--
+-- Uses 'ShortByteString', recommendations and caveats described there apply.
+newtype AtBaseCompact (b :: Symbol) = BaseNShort
+    { fromAtBaseCompact :: ShortByteString
+    } deriving (Eq, Ord, Hashable, NFData)
+
+instance KnownSymbol b => Show (AtBaseCompact b) where
+    showsPrec p (BaseNShort bs) = showParen (p >= 11)
+        ( showString ("Base" <> show (symbolVal (Proxy @b)) <> "Compact ")
+        . showsPrec 11 bs
+        )
+
+compact :: AtBase b -> AtBaseCompact b
+compact = BaseNShort . Short.toShort . fromAtBase
+
+expand :: AtBaseCompact b -> AtBase b
+expand = BaseN . Short.fromShort . fromAtBaseCompact
+
+-- $tagged
+-- 'AtBase' values tagged by the type they're representing.
+
+type Base16Of a = Tagged a (AtBase "16")
+type Base58Of a = Tagged a (AtBase "58")
+type Base64Of a = Tagged a (AtBase "64")
+
+-- $cbor
+-- Directly go from (presumed to be) base-n encoded 'ByteString' to
+-- de-'Serialise'-able value.
+
+data DeserialiseError =
+      DecodeBaseError  String
+    | DeserialiseError DeserialiseFailure
+    deriving Show
+
+deserialiseAtBase
+    :: ( Serialise  a
+       , DecodeBase b
+       )
+    => proxy b
+    -> ByteString
+    -> Either DeserialiseError a
+deserialiseAtBase base bs = do
+    bs' <- bimap DecodeBaseError fromStrict $ decodeAtBaseEither base bs
+    first DeserialiseError $ deserialiseOrFail bs'
+
+-- $encoding
+
+encodeBase16 :: ByteString -> AtBase "16"
+encodeBase16 = BaseN . Base16.encode
+{-# INLINE encodeBase16 #-}
+
+encodeBase64 :: ByteString -> AtBase "64"
+encodeBase64 = BaseN . Base64.encode
+{-# INLINE encodeBase64 #-}
+
+-- |
+-- >>> fromAtBase $ encodeBase16upper "hello world"
+-- "68656C6C6F20776F726C64"
+encodeBase16upper :: ByteString -> AtBase "16u"
+encodeBase16upper = BaseN . C8.map toUpper . Base16.encode
+{-# INLINE encodeBase16upper #-}
+
+-- |
+-- prop> \(Bytes bytes) -> decodeBase16upper (fromAtBase $ encodeBase16upper bytes) === Right bytes
+decodeBase16upper :: ByteString -> Either String ByteString
+decodeBase16upper = decodeBase16Either
+{-# INLINE decodeBase16upper #-}
+
+-- Base 32 ---------------------------------------------------------------------
+
+-- |
+-- fromAtBase . encodeBase32hex $ "hello world"
```

Comment:
Did I write these doctests?

```suggestion
-- >>> fromAtBase . encodeBase32hex $ "hello world"
```

Reply:
I _think_ you did, but I did add the `fromAtBase`. Some of them were imported from `Multibase.hs`, others from `BaseN.hs`, so paternity is unclear 😂